br/pkg/lightning/restore/check_info.go

```diff
@@ -66,252 +46,69 @@ func (rc *Controller) isSourceInLocal() bool {
-	replConfig, err := rc.preInfoGetter.GetReplicationConfig(ctx)
-	if err != nil {
-		return 0, errors.Trace(err)
+// clusterResource check cluster has enough resource to import data. this test can by skipped.
+func (rc *Controller) clusterResource(ctx context.Context) error {
+	checkCtx := ctx
+	if rc.taskMgr != nil {
+		checkCtx = WithPrecheckKey(ctx, taskManagerKey, rc.taskMgr)
 	}
-	return replConfig.MaxReplicas, nil
-}
-
-func (rc *Controller) getClusterAvail(ctx context.Context) (uint64, error) {
-	storeInfo, err := rc.preInfoGetter.GetStorageInfo(ctx)
+	theChecker, err := rc.precheckItemBuilder.BuildPrecheckItem(CheckTargetClusterSize)
 	if err != nil {
-		return 0, errors.Trace(err)
-	}
-	clusterAvail := uint64(0)
-	for _, store := range storeInfo.Stores {
-		clusterAvail += uint64(store.Status.Available)
-	}
-	return clusterAvail, nil
-}
-
-// clusterResource check cluster has enough resource to import data. this test can by skipped.
-func (rc *Controller) clusterResource(ctx context.Context, localSource int64) error {
-	passed := true
-	message := "Cluster resources are rich for this import task"
-	defer func() {
-		rc.checkTemplate.Collect(Critical, passed, message)
-	}()
-
-	var (
-		clusterAvail  uint64
-		clusterSource uint64
-	)
-	if rc.taskMgr == nil {
-		var err error
-		clusterAvail, err = rc.getClusterAvail(ctx)
-		if err != nil {
-			return errors.Trace(err)
-		}
-		clusterSource = uint64(localSource)
-	} else {
-		if err := rc.taskMgr.CheckTasksExclusively(ctx, func(tasks []taskMeta) ([]taskMeta, error) {
-			clusterAvail = 0
-			clusterSource = 0
-			restoreStarted := false
-			for _, task := range tasks {
-				if task.status > taskMetaStatusInitial {
-					restoreStarted = true
-				}
-				clusterSource += task.sourceBytes
-				if task.clusterAvail > 0 {
-					clusterAvail = task.clusterAvail
-				}
-			}
-			if restoreStarted || clusterAvail > 0 {
-				return nil, nil
-			}
-
-			var err error
-			clusterAvail, err = rc.getClusterAvail(ctx)
-			if err != nil {
-				return nil, errors.Trace(err)
-			}
-			newTasks := append([]taskMeta(nil), tasks...)
-			for i := 0; i < len(newTasks); i++ {
-				newTasks[i].clusterAvail = clusterAvail
-			}
-			return newTasks, nil
-		}); err != nil {
-			return errors.Trace(err)
-		}
+		return errors.Trace(err)
 	}
-
-	replicaCount, err := rc.getReplicaCount(ctx)
+	result, err := theChecker.Check(checkCtx)
 	if err != nil {
 		return errors.Trace(err)
 	}
-	estimateSize := clusterSource * replicaCount
-	if estimateSize > clusterAvail {
-		passed = false
-		message = fmt.Sprintf("Cluster doesn't have enough space, available is %s, but we need %s",
-			units.BytesSize(float64(clusterAvail)), units.BytesSize(float64(estimateSize)))
-	} else {
-		message = fmt.Sprintf("Cluster available is rich, available is %s, we need %s",
-			units.BytesSize(float64(clusterAvail)), units.BytesSize(float64(estimateSize)))
+	if result != nil {
```

Comment:
seems only `tableEmptyCheckItem` might return `nil, nil` as return result on context cancelled. 

maybe we can handle that context cancelled from outer logic, then we can can define `Check` as return non null result if err =nil.

Reply:
'nil, nil' means the check is skipped, so there is no check result and no error.  In the code, besides `tableEmptyCheckItem`, `csvHeaderCheckItem` and `checkpointCheckItem` will also return `nil, nil` after figuring out the check doesn't need to be proceeded.  However, I can change the skipped result as a `CheckResult` object with `Passed` setting to `true`, and a message indicating that the check has been skipped.  But IMO it is not necessary to do that.